server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/metrics/TopHitsIT.java

```diff
@@ -1268,6 +1275,44 @@ public void testWithRescore() {
         );
     }
 
+    public void testScriptSorting() {
+        Script script = new Script(ScriptType.INLINE, NAME, "doc['sort'].value", Collections.emptyMap());
+        assertNoFailuresAndResponse(
+            prepareSearch("idx").addAggregation(
+                terms("terms").executionHint(randomExecutionHint())
+                    .field(TERMS_AGGS_FIELD)
+                    .subAggregation(
+                        topHits("hits").sort(SortBuilders.scriptSort(script, ScriptSortType.NUMBER).order(SortOrder.DESC))
+                            .sort(SortBuilders.scoreSort())
+                    )
+            ),
+            response -> {
+                Terms terms = response.getAggregations().get("terms");
+                assertThat(terms, notNullValue());
+                assertThat(terms.getName(), equalTo("terms"));
+                assertThat(terms.getBuckets().size(), equalTo(5));
+
+                double higestSortValue = 0;
+                for (int i = 0; i < 5; i++) {
+                    Terms.Bucket bucket = terms.getBucketByKey("val" + i);
+                    assertThat(bucket, notNullValue());
+                    assertThat(key(bucket), equalTo("val" + i));
+                    assertThat(bucket.getDocCount(), equalTo(10L));
+                    TopHits topHits = bucket.getAggregations().get("hits");
+                    SearchHits hits = topHits.getHits();
+                    assertThat(hits.getTotalHits().value(), equalTo(10L));
+                    assertThat(hits.getHits().length, equalTo(3));
+                    higestSortValue += 10;
+                    assertThat((Double) hits.getAt(0).getSortValues()[0], equalTo(higestSortValue));
+                    assertThat((Double) hits.getAt(1).getSortValues()[0], equalTo(higestSortValue - 1));
+                    assertThat((Double) hits.getAt(2).getSortValues()[0], equalTo(higestSortValue - 2));
+
+                    assertThat(hits.getAt(0).getSourceAsMap().size(), equalTo(5));
+                }
+            }
+        );
+    }
```

Comment:
This test was also missing and if present, would have surfaced concurrency issues when using script sorting as part as top_hits.

Reply:
are you sure that you reverted the changes to ScriptSortBuilder in the right places, still leaving concurrency enabled for script sorting? I get it to fail as I soon as I do that. There is maybe a little less likelihood for it to fail due to randomizing the execution hint, because we parallelize only on global_ordinals execution mode, based on field cardinality. That is for the terms aggs that holds the top_hits agg. But I find that the issue reproduces pretty quickly.

Reply:
Yes, I [reverted all changes](https://github.com/cbuescher/elasticsearch/tree/pr/123757) that are not test-related (tried also with and wothough changes to MockScriptEngine), and ran the test very often from IDE and from command line, no failures so far. I'll try running in a tight loop once again